[RUNTIME] Add cudnn conv3d

optima2005 opened the pull request:
This PR is to add cudnn conv3d in runtime module. And it also modified the existing conv2d to have them use a unified manner to be implemented. This could be the prerequisite of 3D convolution( #4400 ) implementation. 

@masahi

masahi commented on tests/python/contrib/test_cudnn.py:
```diff
@@ -72,6 +72,67 @@ def verify():
 
     verify()
 
```
can you modify this verify() function for cudnn 2D to make sure that output is correct? You can use [conv2d_nchw_python](https://github.com/apache/incubator-tvm/blob/master/topi/python/topi/testing/conv2d_nchw_python.py#L87) function.

masahi commented on tests/python/contrib/test_cudnn.py:
```diff
@@ -72,6 +72,67 @@ def verify():
+    yshape = [x.value for x in Y.shape]
+    s =  tvm.create_schedule(Y.op)
+
+    def verify():
```
Similarly, use conv3d_ncdhw_python that you already have in your earlier PR to verify the output.

optima2005 commented:
I have added the output checking, but it seems there are some accuracy issue. But I didn't find any incorrecion in implementation. Would you check in you platform if convinient? Thanks!

masahi commented:
I also got errors.

For 2D,

```
Mismatch: 0.00229%
Max absolute difference: 1.74964048e-06
Max relative difference: 0.00205616
 x: array([[[[-8.124456e-01,  1.603181e+00, -4.913126e+00, ...,
           3.991851e-01, -1.673781e-01, -1.984869e+00],
         [ 1.048217e+00, -5.826425e-01,  2.729261e+00, ...,...
 y: array([[[[-8.124456e-01,  1.603181e+00, -4.913125e+00, ...,
           3.991851e-01, -1.673780e-01, -1.984869e+00],
         [ 1.048217e+00, -5.826426e-01,  2.729261e+00, ...,...

```

For 3D, 

```
Mismatch: 0.00062%
Max absolute difference: 6.57308036e-06
Max relative difference: 0.4497126
 x: array([[[[[-1.243277e+00,  2.456227e+00, -3.557485e+00, ...,
           -2.807276e+00, -8.497746e-01,  1.374343e+00],
          [-2.806745e+00,  1.923186e+00, -6.345257e+00, ...,...
 y: array([[[[[-1.243278e+00,  2.456226e+00, -3.557484e+00, ...,
           -2.807276e+00, -8.497746e-01,  1.374343e+00],
          [-2.806744e+00,  1.923186e+00, -6.345256e+00, ...,...
```

masahi commented:
But from max abs diff, the result looks correct. cuDNN uses different conv algorithm so some difference in numerical accuracy is expected. I think you can just specify atol=1e-5 for both cases.

You need to do rebase and resolve conflict with the latest master. #4353 modifed cuDNN.

optima2005 commented:
I have rebased to the master

masahi commented:
@optima2005 can you fix the CI problem? Then I can do the review and merge.

optima2005 commented:
@masahi  fixed.

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -59,40 +54,125 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
+
   // Set Filter
-  CUDNN_CALL(cudnnSetFilter4dDescriptor(entry_ptr->conv_entry.filter_desc,
+  for (int i = 0; i < full_dims; i++) {
```
Please fix the indentation of this block

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -59,40 +54,125 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
+      yh = static_cast<int>(y->shape[2]);
+      yw = static_cast<int>(y->shape[3]);
+    }
+
```
I think you can clean up this, Using variables below

batch_index = 0
channel_index = 1
h_index = 2
w_index = 3
if layout = NHWC
   channel_index = 3
   h_index = 1
   w_index = 2

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -120,137 +200,144 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
-                                                   static_cast<int*>(out_shape) + 3));
-});
 
+  if (dims == 2 && entry_ptr->conv_entry.tensor_format ==  CUDNN_TENSOR_NHWC) {
```
why do we need this special case?

optima2005 replied:
without this special case, I always got below error:
```
E             [bt] (2) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(TVMFuncCall+0x90) [0x3fff7599dfa0]
E             [bt] (1) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(+0xeea644) [0x3fff75a5a644]
E             [bt] (0) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(tvm::contrib::OutputShape(int, int, int const*, int const*, int const*, int const*, int const*, void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x1114) [0x3fff75a59a94]
E             File "/host/root/ligc/tmp/incubator-tvm/src/runtime/contrib/cudnn/conv_forward.cc", line 261
E           cuDNN: Check failed: e == CUDNN_STATUS_SUCCESS (3 vs. 0) : CUDNN_STATUS_BAD_PARAM
```
I have double checked the shape for input and weight, they should follow the SDK spec, would you please also take a look to see if any wrong in the code:
```
format

    Input.Type of the filter layout format. If this input is set to CUDNN_TENSOR_NCHW, which is one of the enumerant values allowed by cudnnTensorFormat_t descriptor, then the layout of the filter is as follows:

        For N=4, a 4D filter descriptor, the filter layout is in the form of KCRS:
            K represents the number of output feature maps
            C is the number of input feature maps
            R is the number of rows per filter
            S is the number of columns per filter
        For N=3, a 3D filter descriptor, the number S (number of columns per filter) is omitted.
        For N=5 and greater, the layout of the higher dimensions immediately follow RS.

    On the other hand, if this input is set to CUDNN_TENSOR_NHWC, then the layout of the filter is as follows:

        For N=4, a 4D filter descriptor, the filter layout is in the form of KRSC.
        For N=3, a 3D filter descriptor, the number S (number of columns per filter) is omitted and the layout of C immediately follows R.
        For N=5 and greater, the layout of the higher dimensions are inserted between S and C. For more information, see cudnnTensorFormat_t.
```

masahi commented on tests/python/contrib/test_cudnn.py:
```diff
@@ -67,22 +71,109 @@ def verify_conv2d(data_dtype, conv_dtype, tensor_format=0):
+        if tensor_format == 0:
+            c_np = topi.testing.conv2d_nchw_python(x_np, w_np, 1, 1)
+        elif tensor_format == 1:
+            wt = w_np.transpose((1, 2, 3, 0))  #OHWI => HWIO
```
You don't need this line if you set the wshape at line 52 correctly.

optima2005 replied:
'conv2d_nhwc_python' needs weight layout HWIO, while nudnn needs OHWI.

masahi commented on python/tvm/contrib/cudnn.py:
```diff
@@ -413,3 +411,324 @@ def conv2d_forward(x,
             conv_dtype), name="y")
+
+
+def conv3d_w_shape(in_channel,
```
let's remove this function. This is trivial and only meaningful for NCHW layout. Also remove conv2d_w_shape

masahi commented:
@optima2005 do we still need separate python functions forward, find_algo, output_shape for 2D and 3D cases? We can pass dimension (2d or 3d) and shapes, padding, strides etc by pointers to each functions.

masahi commented:
ok, it doesn't seem obvious how to pass a python list of numbers C++ packed function ...

masahi commented:
@optima2005 please have a look at [this](https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/cudnn.py#L238). It is possible to pass a python list to C++ func by void pointer. 

Can you try unify forward, find_algo, output_shape python functions that we have for both 2D and 3D into one? We can remove all the uglyness.

optima2005 commented:
@masahi ,  could you advice how to write below code to use array?
```
    return _api.extern(
        oshape, [x, w],
        lambda ins, outs: _intrin.call_packed(
            "tvm.contrib.cudnn.conv.forward",
            conv_mode,
            tensor_format,
            algo,
            dims,
            pad,
            stride,
            dilation,
            ins[0],
            ins[1],
            outs[0],
            conv_dtype), name="y")
```
I got this error
```
self = <tvm._ffi.function.Function object at 0x3fff6ebb3240>
args = ('int32', 'tvm_call_packed', ['tvm.contrib.cudnn.conv.forward', 1, 0, 0, 2, [1, 1], ...], 4, None, 0)
temp_args = [["tvm.contrib.cudnn.conv.forward", 1, 0, 0, 2, [1, 1], [1, 1], [1, 1], tvm_stack_make_array(X, tvm_stack_make_shape(3...hape(16, 4, 3, 3), 0, 4, 0f, 0), tvm_stack_make_array(y, tvm_stack_make_shape(3, 16, 32, 32), 0, 4, 0f, 0), "float32"]]
values = <tvm._ffi._ctypes.function.TVMValue_Array_6 object at 0x3fff6adf16a8>
tcodes = <tvm._ffi._ctypes.function.c_int_Array_6 object at 0x3fff6adf1378>

    def __call__(self, *args):
        """Call the function with positional arguments

        args : list
           The positional arguments to the function call.
        """
        temp_args = []
        values, tcodes, num_args = _make_tvm_args(args, temp_args)
        ret_val = TVMValue()
        ret_tcode = ctypes.c_int()
        if _LIB.TVMFuncCall(
                self.handle, values, tcodes, ctypes.c_int(num_args),
                ctypes.byref(ret_val), ctypes.byref(ret_tcode)) != 0:
>           raise get_last_ffi_error()
E           tvm._ffi.base.TVMError: Traceback (most recent call last):
E             [bt] (3) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(TVMFuncCall+0x90) [0x3fff8d50dde0]
E             [bt] (2) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(+0x3cc1bc) [0x3fff8caac1bc]
E             [bt] (1) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(tvm::Array<tvm::Expr, void> tvm::runtime::TVMArgValue::AsObjectRef<tvm::Array<tvm::Expr, void> >() const+0x2ec) [0x3fff8caba96c]
E             [bt] (0) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x60) [0x3fff8ca89960]
E             File "/host/root/ligc/tmp/incubator-tvm/include/tvm/packed_func_ext.h", line 140
E           TVMError: Check failed: ObjectTypeChecker<TObjectRef>: :Check(ptr): Expected type List[Expr] but get Array

python/tvm/_ffi/_ctypes/function.py:207: TVMError
```

masahi commented:
Are you passing python lists? I think you can pass numpy arrays (not list) by void pointer using _get_np_int32_array_handle(). So it should be something like

```
_get_np_int32_array_handle(dims),
_get_np_int32_array_handle(pad),
_get_np_int32_array_handle(stride),
_get_np_int32_array_handle(dilation)
```
assuming dims, pad, etc are numpy array. See [conv2d_output_shape](https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/cudnn.py#L238) for example of using _get_np_int32_array_handle.

masahi commented:
Python arguments that are passed to C++ functions are converted to values that C++ can understand using [this function](https://github.com/apache/incubator-tvm/blob/master/python/tvm/_ffi/_ctypes/function.py#L104).

optima2005 commented:
_get_np_int32_array_handle works for conv_output_shape and conv_find_algo, but not for 'conv_forward', the problem is in the code snippet. If I changed it to
```
    return _api.extern(
        oshape, [x, w],
        lambda ins, outs: _intrin.call_packed(
            "tvm.contrib.cudnn.conv.forward",
            conv_mode,
            tensor_format,
            algo,
            dims,
            _get_np_int32_array_handle(pad),
            _get_np_int32_array_handle(stride),
            _get_np_int32_array_handle(dilation),
            ins[0],
            ins[1],
            outs[0],
            conv_dtype), name="y")

```
where pad, stride, and dilation are np.array(...,  dtype=np.int32), I got this error,
```
value = c_void_p(69268061568800)
    def convert_to_node(value):
        """Convert a python value to corresponding node type.

        ... ...
        ... ...

        if isinstance(value, string_types):
            return _api_internal._str(value)
        if isinstance(value, (list, tuple)):
            value = [convert_to_node(x) for x in value]
            return _api_internal._Array(*value)
        if isinstance(value, dict):
            vlist = []
            for item in value.items():
                if (not isinstance(item[0], _CLASS_NODE_BASE) and
                        not isinstance(item[0], string_types)):
                    raise ValueError("key of map must already been a container type")
                vlist.append(item[0])
                vlist.append(convert_to_node(item[1]))
            return _api_internal._Map(*vlist)
        if isinstance(value, NodeGeneric):
            return value.asnode()
        if value is None:
            return None

>       raise ValueError("don't know how to convert type %s to node" % type(value))
E       ValueError: don't know how to convert type <class 'ctypes.c_void_p'> to node

python/tvm/_ffi/node_generic.py:95: ValueError

```

masahi commented:
Ok, can you push your change to find_algo and output_shape?

optima2005 commented:
I put my changes are in https://github.com/optima2005/incubator-tvm.git (test branch). You can have a try if convenient. Thanks!

masahi commented:
conv_output_shape, conv_find_algo, and conv_forward look good. Once we have them, we don't need conv2d_output_shape, conv3d_output_shape etc that just forward to corresponding generic function are not necessary. Please remove everything in [here](https://github.com/optima2005/incubator-tvm/blob/test/python/tvm/contrib/cudnn.py#L383-L759). 

You can use conv_output_shape, conv_find_algo, and conv_forward from tests and topi. Can you update this PR according to this comment?

masahi commented:
I'll look into how to fix passing python values to call_packed("tvm.contrib.cudnn.conv.forward"). It should be possible.

optima2005 commented:
I remember some tutorials are using those conv2d APIs.  I am wondering if we should keep them for a while and mark them deprecated.

masahi commented:
Right I wrote that tutorial :) cuDNN is used via topi, so you need to update [this line](https://github.com/apache/incubator-tvm/blob/master/topi/python/topi/cuda/conv2d.py#L99). That one and the test cases in test_cudnn.py are the only place that use cudnn, so as long as you update them it is fine.

optima2005 commented:
ok

masahi commented:
@optima2005 Can you also quickly try using [tvm.api.convert()](https://github.com/apache/incubator-tvm/blob/master/python/tvm/api.py#L121) function in call_packed? We are using to pass a python list of strings to C++, for example. It is used [here](https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/rocm.py#L139) and on the C++ side [here](https://github.com/apache/incubator-tvm/blob/master/src/codegen/llvm/codegen_amdgpu.cc#L232-L234).

I want to know if it is possible to pass arguments in call_packed via convert([pad_h, pad_w]) ... etc.

optima2005 commented:
I may not have time now. I will look it tomorrow.

masahi commented:
@optima2005 that's all right. I made a bunch of minor comments regarding code style. Please do not pass dim explicitly and keep std::vector around as much as possible.

masahi commented on src/runtime/contrib/cudnn/cudnn_utils.h:
```diff
@@ -54,6 +54,13 @@ inline void GetStride(int nbdim, const int *dims, int *strides) {
   }
 }
 
+inline void GetCudnnStride(int nbdim, const int *dims, int *strides) {
```
Use std::vector and pass by reference

masahi commented on topi/python/topi/testing/conv3d_ncdhw_python.py:
```diff
@@ -0,0 +1,127 @@
+
+
+def conv3d_ncdhw_python(a_np, w_np, stride, padding, groups=1):
+    """Convolution operator in NCHW layout.
```
NCDHW

masahi commented on topi/python/topi/testing/conv3d_ncdhw_python.py:
```diff
@@ -0,0 +1,127 @@
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, line-too-long, unused-variable, too-many-locals, too-many-branches
+"""Convolution in python"""
```
Convolution 3D

masahi commented on topi/python/topi/testing/conv3d_ncdhw_python.py:
```diff
@@ -0,0 +1,127 @@
+
+
+def _conv3d_ncdhw_python(a_np, w_np, stride, padding):
+    """Convolution operator in NCDHW layout.
```
This is an internal function, so no need for docstring

masahi commented on tests/python/contrib/test_cudnn.py:
```diff
@@ -37,52 +38,127 @@ def verify_conv2d(data_dtype, conv_dtype, tensor_format=0):
+    batch = 3
+    depth = 32
+    height = 32
+    weight = 32
```
Clean these up using list

Also do the same for conv2d test.

optima2005 commented:
@masahi , Thanks for the comments! I will revise them after I have fixed the pass array into C issue.
I changed to use  tvm.api.convert() solution, but still get below error:
```
self = <tvm._ffi.function.Function object at 0x3fff6d193318>
args = ('int32', 'tvm_call_packed', ['tvm.contrib.cudnn.conv.forward', 1, 0, 0, 2, [1, 1], ...], 4, None, 0)
temp_args = [["tvm.contrib.cudnn.conv.forward", 1, 0, 0, 2, [1, 1], [1, 1], [1, 1], tvm_stack_make_array(X, tvm_stack_make_shape(3...hape(16, 4, 3, 3), 0, 4, 0f, 0), tvm_stack_make_array(y, tvm_stack_make_shape(3, 16, 32, 32), 0, 4, 0f, 0), "float32"]]
values = <tvm._ffi._ctypes.function.TVMValue_Array_6 object at 0x3fff693b8d90>
tcodes = <tvm._ffi._ctypes.function.c_int_Array_6 object at 0x3fff693b8f28>

    def __call__(self, *args):
        """Call the function with positional arguments

        args : list
           The positional arguments to the function call.
        """
        temp_args = []
        values, tcodes, num_args = _make_tvm_args(args, temp_args)
        ret_val = TVMValue()
        ret_tcode = ctypes.c_int()
        if _LIB.TVMFuncCall(
                self.handle, values, tcodes, ctypes.c_int(num_args),
                ctypes.byref(ret_val), ctypes.byref(ret_tcode)) != 0:
>           raise get_last_ffi_error()
E           tvm._ffi.base.TVMError: Traceback (most recent call last):
E             [bt] (3) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(TVMFuncCall+0x90) [0x3fff8bd8de00]
E             [bt] (2) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(+0x3cc1dc) [0x3fff8b32c1dc]
E             [bt] (1) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(tvm::Array<tvm::Expr, void> tvm::runtime::TVMArgValue::AsObjectRef<tvm::Array<tvm::Expr, void> >() const+0x2ec) [0x3fff8b33a98c]
E             [bt] (0) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x60) [0x3fff8b309980]
E             File "/host/root/ligc/tmp/incubator-tvm/include/tvm/packed_func_ext.h", line 140
E           TVMError: Check failed: ObjectTypeChecker<TObjectRef>: :Check(ptr): Expected type List[Expr] but get Array

python/tvm/_ffi/_ctypes/function.py:207: TVMError
```
I put the changes [here](https://github.com/optima2005/incubator-tvm/commit/983433fe7787a45b2a0994eebe5b3f8f2e1970aa), could you please provide advice?

masahi commented:
@tqchen @vinx13 Why can't we pass a python list of ints to intrin.call_packed(...) using tvm.api.convert()? We are passing a python list as Array<Expr> but TVM expects List[Expr] type?

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -22,31 +22,29 @@
+  int mode,
+  int format,
+  int algo,
+  const std::vector<int> pad,
```
pass by reference

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -120,137 +180,145 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
+  const std::vector<int> stride,
+  const std::vector<int> dilation,
+  const std::vector<int> x_dim,
+  const std::vector<int> w_dim,
```
pass by reference

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -120,137 +180,145 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
+  const std::vector<int> dilation,
+  const std::vector<int> x_dim,
+  const std::vector<int> w_dim,
+  const std::vector<int> y_dim,
```
pass by reference

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -287,6 +355,91 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.find_algo")
+  Array<Expr> pads =  args[4];
+  Array<Expr> strides =  args[5];
+  Array<Expr> dilations =  args[6];
+
```
This seems wrong... Expr and Array are compile-time stuff, they should not appear in the runtime code.
Ideally they should be void or int pointers, but I don't know how to make it work.

optima2005 replied:
How about form a string on python and parse it in C? But only if we don't have a better approach?

masahi replied:
No that sounds bad. In the worst case we need to go back to explicitly passing all values (i.e. we need to have separate conv2d_forward and conv3d_forward)

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -287,6 +355,91 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.find_algo")
+  std::vector<int> dilation_v(dilation, dilation + dims);
+  std::vector<int> x_dim_v(x_dim, x_dim + dims + 2);
+  std::vector<int> w_dim_v(w_dim, w_dim + dims + 2);
+
```
These vectors don't seem necessary. I told you to pass arguments by vector yesterday, but now that I realized we don't need vectors in the first place. Sorry about that.

Please remove these vectors.

optima2005 replied:
why? Then how those tuples should be passed into the OutputShape(), FindAlgo()...?

masahi replied:
Sorry for the confusion. Originally (until yesterday) those functions take dim and pointers as argument right? But you were also using std::vector, so I thought why don't we use vectors as much as possible.

But now that I realized that those vectors were just some placeholders and doesn't add anything, I think we should remove vectors and restore your original implementation.

optima2005 replied:
Originally I only use vector to handle the different tuple size for 2d and 3d.  After we have unified all those 2d, 3d stuff, they replace most of the C style arrays.

optima2005 replied:
Because most need handle the different tuple size, I am wondering whether we need keep them.

masahi replied:
All you do with these vectors in OutputShape and FindAlgo is just .data() right? So why not pass pad, stride (of int*) etc directly rather than pad_v, stride_v (you need to update OutputShape and FindAlgo, of course). Is there something you can only do with vectors?

masahi replied:
ping @optima2005

masahi commented on src/runtime/contrib/cudnn/conv_forward.cc:
```diff
@@ -287,6 +355,91 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.find_algo")
+  std::vector<int> dilation_v(dilation, dilation + dims);
+  std::vector<int> x_dim_v(x_dim, x_dim + dims + 2);
+  std::vector<int> w_dim_v(w_dim, w_dim + dims + 2);
+  std::vector<int> y_dim_v(y_dim, y_dim + dims + 2);
```
remove vectors

masahi commented:
@optima2005 The problem seems to be that "Array" is not an "Expr". make.Call [expects](https://github.com/apache/incubator-tvm/blob/master/src/api/api_ir.cc#L102) its argument to be "Array\<Expr\>", but if we pass a python list, which converts to "Array", TVM complains that "Array" is not an "Expr". The error message is confusing, but that actual error occurs [here](https://github.com/apache/incubator-tvm/blob/master/include/tvm/packed_func_ext.h#L70), where TVM tries to type check Array elements type. The element type should be "Expr", but if we pass a list as an additional argument, the type checking fails because it is "Array", not "Expr". 

The reason we can pass Tensors is that, before TVM passes python arguments to C++,  TVM replaces Tensors with "a Call expression that returns Tensor" [(here)](https://github.com/apache/incubator-tvm/blob/master/python/tvm/intrin.py#L41-L42). We need something similar for lists, but that doesn't seem to exist now.

So we need to give up on our approach. Can you add python functions conv2d_forward and conv3d_forward? Once the CI passes, I'll check the std::vector issue and then merge.

optima2005 commented:
@masahi , I add back conv2d and 3d. Please take a look if it is expected.
For vector issue, another approach I can think is new array[], I will follow you decision.

optima2005 commented:
BTW. Many thanks for the detail explanation about how those Exp, Array... in tvm works. I haven't full understand them yet. I think I need a little more time to learn tvm internal working mechanism.

masahi commented on src/runtime/contrib/cudnn/cudnn_utils.h:
```diff
@@ -54,6 +55,15 @@ inline void GetStride(int nbdim, const int *dims, int *strides) {
 
+inline void GetCudnnStride(int nbdim,
+                           const std::vector<int>& dims,
+                           std::vector<int>* strides) {
```
std::vector<int>& strides

optima2005 replied:
If using reference, it would cause lint failure: "Is this a non-const reference? If so, make const or use a pointer: std::vector<int>& strides  [runtime/references] [2]"

masahi replied:
ok, didn't know the linter favors such style.

optima2005 commented:
Just found some headers are unused, removed them.

masahi commented:
@optima2005 can you address my comments on the std::vector issue? I still think these vectors are not necessary, as long as we have a pointer and dimension.

optima2005 commented:
@masahi, ok. i will remove them. I will use a new array[] to handle the various size of array for 2d and 3d.

masahi commented:
no that's not what I meant. Why do you need to allocate another array (either vector or new[])? Am I missing something?

masahi commented:
@optima2005 Have you read my comment [here](https://github.com/apache/incubator-tvm/pull/4418/files#r352275702)?

optima2005 commented:
The size of pad, stride and dilation tuple is 2 for 2d, and 3 for 3d. So those vectors can act as kind of 'array with variable initial size'. This is not supported by c++ directly.  This is the only reason I am using them.

masahi commented:
My point is, why do you need array with variable initial size in the first place? You can just pass pointers directly

optima2005 commented:
I got it.  Sorry for confused!

masahi commented:
@optima2005 thanks for being patient with me:) Let's continue with your original PR.
